### PR TITLE
Order users in autocomplete by length

### PR DESCRIPTION
**Pull request by SimonBrandner**
Fixes https://github.com/vector-im/element-web/issues/15609

<hr>

Screenshot_20210711_084053

<hr>

This isn't an ideal solution since it doesn't take the server part of `mxid`s into account when searching for the reasons explained in https://github.com/vector-im/element-web/issues/15609#issuecomment-877690692

### CODE REVIEW

File: src/autocomplete/UserProvider.tsx [hunk 1]
```diff
@@ -55,7 +55,7 @@ export default class UserProvider extends AutocompleteProvider {
         this.room = room;
         this.matcher = new QueryMatcher([], {
             keys: ['name'],
-            funcs: [obj => obj.userId.slice(1)], // index by user id minus the leading '@'
+            funcs: [obj => obj.userId.slice(1).split(":")[0]], // index by user id minus the leading '@' and the serer part
```
Comment (turt2live) [inline src/autocomplete/UserProvider.tsx:58]:
```suggestion
            funcs: [obj => obj.userId.slice(1).split(":")[0]], // index by user id minus the leading '@' and the server part
```

**Review by turt2live — APPROVED** [review]
the code looks fine, though personally I'd prefer a relevancy metric over a length metric. If I share a lot of rooms with Matthew, and ping Matthew a lot, and talk to Matthew a lot, I probably want to ping that Matthew.

**Comment by MadLittleMods** [comment]
This seems like more of a product decision vs a design thing. Switched the reviewers ⏩

**Comment by nadonomy** [comment]
@SimonBrandner I know this has been a while but I think this is a risky change so we should close/not merge this.

I can see the attraction of exact matching if searching for something short, but we'll break peoples existing muscle memory so it's a really high risk change. I also agree with [Travis's comment](https://github.com/matrix-org/matrix-react-sdk/pull/6348#pullrequestreview-712313137) that we should figure out some smarter relevancy metrics instead, perhaps informed by space membership too.

Sorry to be the bearer of bad news!

**Comment by SimonBrandner** [comment]
No worries, makes sense, could you please comment/close the issue?

**Pull request closed without merging**